sympy/simplify/cse_main.py

```diff
@@ -714,7 +714,7 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
     temp = []
     for e in exprs:
         if isinstance(e, (Matrix, ImmutableMatrix)):
-            temp.append(Tuple(*e._mat))
+            temp.append(Tuple(*e._flat))
```

Comment:
Perhaps this attribute shouldn't be used at all outside of the matrices submodule. `e[:]` appears to give the same thing.

Reply:
Agreed, I've changed it to that.

It should be possible to use just `Tuple(*e)` here but that's very slow for large matrices. Apparently `iter(Matrix)` uses the `__len__/__getitem__` version of the iterator protocol which is slow because `__getitem__` is slow. It would be better to add something like:
```python
def __iter__(self):
    return iter(self[:])  # return a list iterator
```
I'm not even sure what class things like that belong in though because there are so many different matrix classes.